Load R2R images on osx-arm64 according to Apple's guidelines

jgiannuzzi opened the pull request:
Closes #65341.

This PR changes how R2R images are loaded on Apple Silicon by following [Apple's guidelines](https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon#Disable-Write-Protections-Before-You-Generate-Instructions). Without this change, debugging is [broken](https://github.com/dotnet/runtime/issues/65341) on this platform, as the R2R memory cannot be written to.

jgiannuzzi committed:
Load R2R images on osx-arm64 according to Apple's guidelines

jgiannuzzi commented:
@janvorli @hoyosjs please take a look!

janvorli commented on src/coreclr/vm/peimagelayout.cpp:
```diff
@@ -223,6 +223,7 @@ void PEImageLayout::ApplyBaseRelocations(bool relocationMustWriteCopy)
 
         BYTE * pageAddress = (BYTE *)GetBase() + rva;
 
+#if !defined(__APPLE__) && !defined(HOST_ARM64)
```
I think this condition is incorrect and it should be:
```suggestion
#if !(defined(__APPLE__) && defined(HOST_ARM64))
```

jgiannuzzi replied:
agreed, thanks!

VSadov commented on src/coreclr/pal/src/map/map.cpp:
```diff
@@ -2041,14 +2041,33 @@ MAPmmapAndRecord(
+        // Set the requested mapping with forced PROT_WRITE, mmap the file, and copy its contents there.
+        // PROT_WRITE can then be removed on Intel, but not on Apple Silicon which requires the use of
+        // pthread_jit_write_protect to switch between executable and writable.
+        LPVOID pvMappedFile = mmap(NULL, len + adjust, PROT_READ, MAP_SHARED, fd, offset - adjust);
```
I think it is aligned. There is assert just above:

```
_ASSERTE(OffsetWithinPage(offset - adjust) == 0);
```

janvorli replied:
Ok, great then. I guess it was just some old time issue that I've remembered from the discussions.

VSadov replied:
I vaguely remember adding the asserts, but do not remember when.

VSadov replied:
Added in https://github.com/dotnet/runtime/pull/39454

VSadov commented on src/coreclr/pal/src/map/map.cpp:
```diff
@@ -2041,14 +2041,33 @@ MAPmmapAndRecord(
+        // Set the requested mapping with forced PROT_WRITE, mmap the file, and copy its contents there.
+        // PROT_WRITE can then be removed on Intel, but not on Apple Silicon which requires the use of
+        // pthread_jit_write_protect to switch between executable and writable.
+        LPVOID pvMappedFile = mmap(NULL, len + adjust, PROT_READ, MAP_SHARED, fd, offset - adjust);
```
Do we need `MAP_SHARED` here?

VSadov replied:
Since this is a sequential read, maybe just use `pread` ?

janvorli replied:
`pread` doesn't work when attempting to read into a memory mapped with MAP_JIT flag.

janvorli replied:
`MAP_PRIVATE` would be better here.

jgiannuzzi replied:
agreed!

jkotas commented on src/coreclr/vm/peimagelayout.cpp:
```diff
@@ -279,7 +281,14 @@ void PEImageLayout::ApplyBaseRelocations(bool relocationMustWriteCopy)
             case IMAGE_REL_BASED_PTR:
+#if defined(__APPLE__) && defined(HOST_ARM64)
+                {
+                    ExecutableWriterHolder<TADDR> addressWriterHolder((TADDR *) address, sizeof(TADDR));
```
It looks very expensive performance-wise to take the writer holder for every reloc.

janvorli replied:
Please note that this is only for arm64 macOS, so this is cheap as it only calls the API to switch the current CPU mode. However, it would probably be better to use the holder around the whole function or call the PAL_JitWriteProtect at the beginning and end of this function to be more explicit with the fact that this is apple specific.

jgiannuzzi replied:
You're right — I did a quick calculation and it's about 9 times slower in average on applying relocations on the images used by the basic console template. Though we are talking really small numbers: 140us vs 940us for all the relocations on a M1 Max. I have a version where I take the writer holder once per section and it goes down to 100us on the same machine. The code is slightly more complex and harder to follow, but I'm happy to commit it to this branch if you'd like.

jgiannuzzi replied:
I have reimplemented it with @janvorli's suggestion of using `PAL_JitWriteProtect` at the beginning and end of `PEImageLayout::ApplyBaseRelocations`. It's as simple and readable as my original change, and much faster.
Thanks to both Jans!

jgiannuzzi committed:
Address review comments

* use MAP_PRIVATE for the mapped file
* correct the condition

jgiannuzzi committed:
Use PAL_JitWriteProtect once per relocation

jgiannuzzi committed:
Flush instruction cache during relocation on Apple Silicon

jgiannuzzi committed:
Fix typo in comment

jgiannuzzi commented on src/coreclr/vm/peimagelayout.cpp:
```diff
@@ -299,8 +306,12 @@ void PEImageLayout::ApplyBaseRelocations(bool relocationMustWriteCopy)
             }
         }
 
+#if defined(__APPLE__) && defined(HOST_ARM64)
+        BOOL bExecRegion = true;
+#else
         BOOL bExecRegion = (dwOldProtection & (PAGE_EXECUTE | PAGE_EXECUTE_READ |
             PAGE_EXECUTE_READWRITE | PAGE_EXECUTE_WRITECOPY)) != 0;
+#endif
```
I have added this to make sure that the instruction cache gets flushed on Apple Silicon too. I am not sure whether this is required?

jgiannuzzi replied:
Agreed. I changed the logic to use `PAL_JitWriteProtect` where the protection is toggled on other platforms to minimise range to flush.

jgiannuzzi committed:
Use PAL_JitWriteProtect in the same spots where protection is toggled

This allows to keep the instruction cache flush logic identical between platforms.

hoyosjs reviewed (approved):
Looks good to me now. Not sure if pulsing the write mode is a perf concern as before. Seems in line with what we did before and seems pretty clean now in terms of readability. Also confirmed that the debugger breakpoints work as expected and that the code that's running comes from the R2R image itself.﻿ Thanks @jgiannuzzi

jgiannuzzi commented on src/coreclr/pal/src/map/map.cpp:
```diff
@@ -2041,14 +2041,33 @@ MAPmmapAndRecord(
-            (pread(fd, pvBaseAddress, len + adjust, offset - adjust) == -1) ||
-            (mprotect(pvBaseAddress, len + adjust, prot) == -1))
+        // Set the requested mapping with forced PROT_WRITE, mmap the file, and copy its contents there.
+        // PROT_WRITE can then be removed on Intel, but not on Apple Silicon which requires the use of
```
It's not needed on Intel — doing a `pread` like before works. I have changed it like this to unify as much as possible between the two architectures.

jgiannuzzi replied:
alright — I will rework this part to keep it the same on Intel

jgiannuzzi replied:
Done in a1d528d.

jgiannuzzi committed:
Revert the change for Intel to avoid any regression

jgiannuzzi commented on src/coreclr/vm/peimagelayout.cpp:
```diff
@@ -251,18 +259,21 @@ void PEImageLayout::ApplyBaseRelocations(bool relocationMustWriteCopy)
                 if (((pSection->Characteristics & VAL32(IMAGE_SCN_MEM_EXECUTE)) != 0))
                 {
-#ifdef __APPLE__
-                    dwNewProtection = PAGE_READWRITE;
```
@jkotas Note that this was introduced in #61938 for fixing the loading of R2R images on Apple Silicon. I believe this was never needed for Intel, hence why I removed it here. @VSadov could you please confirm that this sounds right to you?

jgiannuzzi committed:
Always unmap the file to avoid leaks + log errors

jgiannuzzi commented on src/coreclr/pal/src/map/map.cpp:
```diff
@@ -2041,14 +2041,38 @@ MAPmmapAndRecord(
+        if (pvMappedFile == MAP_FAILED ||
+            (mprotect(pvBaseAddress, len + adjust, prot | PROT_WRITE) == -1))
+        {
+            palError = FILEGetLastErrorFromErrno();
```
addressed in 5d8b591.

janvorli reviewed (approved):
LGTM, thank you!

mangod9 commented:
Thanks for fixing @jgiannuzzi

mangod9 commented:
/backport to release/6.0

jgiannuzzi commented:
@mangod9 I can take care of the backport — it will have to include my reverted PR #64104 as well to be effective.
